### PR TITLE
Fix spec compliance for `String#<<`

### PR DESCRIPTION
**Pull request by lopopolo**
Fix issues with `String#<<`:

- Raise `FrozenError` if the receiver is frozen.
- Add support for pushing codepoints to a string.
- Make pushing codepoints to a string encoding-aware.
- Add support for in-place encoding mutations when pushing non-ASCII bytes to an ASCII-encoded string.
- Make pushing strings to a string encoding-aware.
- Add support for in-place encoding mutations when pushing to an empty string.

Progress on https://github.com/artichoke/artichoke/issues/1912.

To make these changes, some additional feature work was required:

- Add a `spinoso_string::String::set_encoding` API (@b-n: it's back!).
- Fix a bug in `impl BoxUnboxVmValue for String` where encoding flags were not being zeroed before repacking.
- Add a new implicit conversion routine to string where the underlying `spinoso_string::String` is returned instead of the underlying bytes.

## ruby/spec

Before: Passed 1917, skipped 250, not implemented 15, failed 0 specs.
After: Passed 1932, skipped 260, not implemented 15, failed 2 specs.

I declined to add any additional tests since all of this new feature work is exercised by ruby/spec. After these changes, these are the only failures (due to missing support for `String#encode` and `Encoding::UTF16_LE`):

```
Passed 1932, skipped 260, not implemented 15, failed 2 specs.

String#<< when self is in an ASCII-incompatible encoding incompatible with the argument's encoding raises Encoding::CompatibilityError if neither are empty
SpecExpectationNotMetError: Expected Encoding::CompatibilityError
but no exception was raised ("xy" was returned)

/artichoke/virtual_root/src/lib/core/string/shared/concat.rb:115:in protect
(eval):165:in all?
(eval):590:in each
(eval):590:in each
/artichoke/virtual_root/src/lib/core/string/append_spec.rb:8
(eval):590:in each
/artichoke/virtual_root/src/lib/spec_runner.rb:74:in run_specs
(eval):1

String#<< when the argument is in an ASCII-incompatible encoding incompatible with self's encoding raises Encoding::CompatibilityError if neither are empty
SpecExpectationNotMetError: Expected Encoding::CompatibilityError
but no exception was raised ("xy" was returned)

/artichoke/virtual_root/src/lib/core/string/shared/concat.rb:133:in protect
(eval):165:in all?
(eval):590:in each
(eval):590:in each
/artichoke/virtual_root/src/lib/core/string/append_spec.rb:8
(eval):590:in each
/artichoke/virtual_root/src/lib/spec_runner.rb:74:in run_specs
(eval):1

Passed 1932, skipped 260, not implemented 15, failed 2 specs.
```

### CODE REVIEW

File: spinoso-string/src/enc/mod.rs
Comment (lopopolo) [inline spinoso-string/src/enc/mod.rs]: this will need a rebase once #1976 lands.

Comment (lopopolo) [reply spinoso-string/src/enc/mod.rs]: done

File: artichoke-backend/src/convert/implicit.rs
Comment (b-n) [inline artichoke-backend/src/convert/implicit.rs:563]: What are your thoughts on making this a `implicitly_convert_to_nilable_spinoso_string` and making it return `Result<Option<UnboxedValueGuard<'a, spinoso_string::String>>, Error>` ?

I'd imagine this might enable to checking of things like this easier:

```
irb(main):001:0> Time.at(1000, in: "")
Traceback (most recent call last):
        5: from /home/ben/.rbenv/versions/2.6.3/bin/irb:23:in `<main>'
        4: from /home/ben/.rbenv/versions/2.6.3/bin/irb:23:in `load'
        3: from /home/ben/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/irb-1.0.0/exe/irb:11:in `<top (required)>'
        2: from (irb):1
        1: from (irb):1:in `at'
ArgumentError ("+HH:MM" or "-HH:MM" expected for utc_offset)
irb(main):002:0> Time.at(1000, in: nil)
Traceback (most recent call last):
        6: from /home/ben/.rbenv/versions/2.6.3/bin/irb:23:in `<main>'
        5: from /home/ben/.rbenv/versions/2.6.3/bin/irb:23:in `load'
        4: from /home/ben/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/irb-1.0.0/exe/irb:11:in `<top (required)>'
        3: from (irb):2
        2: from (irb):2:in `rescue in irb_binding'
        1: from (irb):2:in `at'
TypeError (can't convert nil into an exact number)
irb(main):003:0> 
```

:point_up: There might be better ways to implement this kind of switching, since in theory `Time.at(1000, in: 1000)` is also valid (e.g. the switch logic needs to handle 3 types in this above case :thinking:)

I can imagine the moot of `nil` is never a `String`, so your current implementation feels more correct, but just thinking out loud.

Comment (b-n) [inline artichoke-backend/src/convert/implicit.rs:563]: Additionally, another alternative `implicitly_convert_to_ruby_string` ?

File: artichoke-backend/src/extn/core/string/trampoline.rs
Comment (b-n) [inline artichoke-backend/src/extn/core/string/trampoline.rs:95]: Does this really need the conversion?

```
irb(main):001:0> a = "hello"
=> "hello"
irb(main):002:0> i = 64
=> 64
irb(main):003:0> a << i
=> "hello@"
irb(main):004:0> a.encoding
=> #<Encoding:UTF-8>
irb(main):005:0> a.force_encoding(Encoding::ASCII)
=> "hello@"
irb(main):006:0> a.encoding
=> #<Encoding:US-ASCII>
irb(main):007:0> a << i
=> "hello@@"
irb(main):008:0> a.encoding
=> #<Encoding:US-ASCII>
```

It looks like it survives as ASCII afterwards

File: artichoke-backend/src/extn/core/string/trampoline.rs
Comment (b-n) [inline artichoke-backend/src/extn/core/string/trampoline.rs:76]: Rust is crazy cool with it's conversions here. I was just playing with `try_push_codepoint` expecting it to not work, but of course it does work. I will play with this one day, since `char.try_from(192)` giving unicode `\u00C0` is very cool and partially black magic to me right now.

File: artichoke-backend/src/extn/core/string/trampoline.rs
Comment (b-n) [inline artichoke-backend/src/extn/core/string/trampoline.rs:64]: How do you feel about `push_int` in `spinoso-string` and letting the encoded types to deal with this?

File: spinoso-string/src/enc/mod.rs
Comment (b-n) [inline spinoso-string/src/enc/mod.rs:159]: Nice!

Just checking, `mem::take` is letting this function own self, and then we're just setting that ownership back to self with the `*self = ` assignment right? That's very handy in this niche cache.

File: artichoke-backend/src/convert/implicit.rs
Comment (lopopolo) [reply artichoke-backend/src/convert/implicit.rs:563]: We have both `implicitly_convert_to_string` and `implicitly_convert_to_nilable_string`. If you find a need for implicitly converting to a nilable spinoso string, it is just a few lines to add that function. Since I don't need it here, I'd prefer to hold off on adding it.

https://github.com/artichoke/artichoke/blob/9a979d9f57ad2afe4b7d312e91a07e15aee93d28/artichoke-backend/src/convert/implicit.rs#L475-L485

File: artichoke-backend/src/extn/core/string/trampoline.rs
Comment (lopopolo) [reply artichoke-backend/src/extn/core/string/trampoline.rs:76]: 🦀 🚀

File: artichoke-backend/src/extn/core/string/trampoline.rs
Comment (lopopolo) [reply artichoke-backend/src/extn/core/string/trampoline.rs:95]: 😉 try this again with `s << 255`. That's what the `!byte.is_ascii()` checks for.

File: spinoso-string/src/enc/mod.rs
Comment (lopopolo) [reply spinoso-string/src/enc/mod.rs:159]: `mem::take` moves the `EncodedString` struct out of `&mut self` and leaves `EncodedString::default()` in its place.

Then we grab its buf, construct a new `EncodedString` with the buf and given encoding.

Then we move that new `EncodedString` back into the `&mut self`.

throughout this whole dance, `self` is valid and initialized.

File: artichoke-backend/src/convert/implicit.rs
Comment (lopopolo) [reply artichoke-backend/src/convert/implicit.rs:563]: /shrug I'm ambivalent between both of these options.

File: artichoke-backend/src/extn/core/string/trampoline.rs
Comment (lopopolo) [reply artichoke-backend/src/extn/core/string/trampoline.rs:64]: that's a good idea! let me file a ticket.

Edit: done!

- https://github.com/artichoke/artichoke/issues/1981

File: artichoke-backend/src/convert/implicit.rs
Comment (lopopolo) [reply artichoke-backend/src/convert/implicit.rs:563]: > the switch logic needs to handle 3 types in this above case

One thing that I think is a constant across all of the type conversions is that only one argument shape gets the implicit conversion treatment. So a method will either implicitly convert to string OR int, but not both.

That implicit conversion needs to go last in the chain so we can propagate the `TypeError`s.

Comment (b-n) [reply artichoke-backend/src/convert/implicit.rs:563]: > Since I don't need it here, I'd prefer to hold off on adding it.

Yes good call!

Comment (b-n) [reply artichoke-backend/src/convert/implicit.rs:563]: Likewise! Resolving :-D

File: artichoke-backend/src/extn/core/string/trampoline.rs
Comment (b-n) [reply artichoke-backend/src/extn/core/string/trampoline.rs:95]: hah oof :grimacing:

**Review by b-n — APPROVED** [review]
LGTM!

**Pull request merged**
